File: cpp/src/isolated_aggregate.cpp
#include "tide/isolated_aggregate.h"
#include <torch/csrc/autograd/custom_function.h>
#include <stdexcept>

namespace tide {
namespace {
using torch::autograd::AutogradContext;
using torch::autograd::variable_list;
class IsolatedAggregate final : public torch::autograd::Function<IsolatedAggregate> {
 public:
  static variable_list forward(AutogradContext* ctx, at::TensorList atoms,
      at::TensorList scales, Tensor coefficients, Index sources, bool mean) {
    ctx->set_materialize_grads(false);
    const Index rows = atoms.size()/sources;
    auto x = at::stack(atoms).reshape({rows, sources, -1});
    auto s = at::stack(scales).reshape({rows, sources});
    auto values = x*s.unsqueeze(-1);
    if (mean) values = values/static_cast<double>(sources);
    else if (coefficients.numel()) values = values*coefficients.reshape({coefficients.dim() == 2 ? rows : 1, sources, 1});
    // Match the scalar program's source-order addition, not a different reduction.
    auto total = values.select(1, 0);
    for (Index j = 1; j < sources; ++j) total = total+values.select(1, j);
    variable_list outputs, frozen;
    for (Index i = 0; i < rows; ++i) {
      auto summary = total[i]; outputs.push_back(summary);
      bool active = false;
      for (Index j = 0; j < sources; ++j) {
        const auto k = i*sources+j;
        auto value = values[i][j]; outputs.push_back(value);
        const bool needs = atoms[k].requires_grad() || scales[k].requires_grad() || coefficients.requires_grad();
        active |= needs;
        if (!needs) frozen.push_back(value);
      }
      if (!active) frozen.push_back(summary);
    }
    ctx->mark_non_differentiable(frozen);
    ctx->saved_data["sources"] = sources;
    ctx->saved_data["mean"] = mean;
    // Packing owns snapshots, like stack followed by elementwise multiplication.
    ctx->save_for_backward({x, s, coefficients});
    return outputs;
  }
  static variable_list backward(AutogradContext* ctx, variable_list grads) {
    if (at::GradMode::is_enabled()) throw std::invalid_argument("isolated Aggregate supports first-order VJP only");
    const auto sources = ctx->saved_data["sources"].toInt();
    const bool mean = ctx->saved_data["mean"].toBool();
    const Index rows = grads.size()/(sources+1), count = rows*sources;
    variable_list result(2*count+3);
    const auto saved = ctx->get_saved_variables();
    const auto& x = saved[0]; const auto& s = saved[1]; const auto& coe = saved[2];
    Tensor dc;
    for (Index j = 0; j < sources; ++j) {
      std::vector<Index> used;
      variable_list bars;
      bool need_x = false, need_s = false;
      for (Index i = 0; i < rows; ++i) {
        const auto& total = grads[i*(sources+1)];
        const auto& part = grads[i*(sources+1)+1+j];
        if (!total.defined() && !part.defined()) continue;
        used.push_back(i);
        bars.push_back(total.defined() && part.defined() ? total+part : total.defined() ? total : part);
        need_x |= ctx->needs_input_grad(i*sources+j);
        need_s |= ctx->needs_input_grad(count+i*sources+j);
      }
      if (used.empty()) continue; // Numerical zero is still a connected cotangent.
      auto ids = at::tensor(used, x.options().dtype(at::kLong));
      auto dy = at::stack(bars);
      auto xs = x.select(1,j).index_select(0,ids);
      auto ss = s.select(1,j).index_select(0,ids);
      auto dv = dy;
      if (mean) dv = dy/static_cast<double>(sources);
      else if (coe.numel()) dv = dy*(coe.dim() == 2 ? coe.index_select(0,ids).select(1,j).unsqueeze(-1) : coe[j]);
      if (need_x) {
        auto dx = dv*ss.unsqueeze(-1);
        for (size_t k = 0; k < used.size(); ++k)
          if (ctx->needs_input_grad(used[k]*sources+j)) result[used[k]*sources+j] = dx[k];
      }
      if (need_s) {
        auto ds = (dv*xs).sum(-1);
        for (size_t k = 0; k < used.size(); ++k)
          if (ctx->needs_input_grad(count+used[k]*sources+j)) result[count+used[k]*sources+j] = ds[k];
      }
      if (coe.numel() && ctx->needs_input_grad(2*count)) {
        if (!dc.defined()) dc = at::zeros_like(coe);
        auto terms = dy*(xs*ss.unsqueeze(-1));
        if (coe.dim() == 2) dc.select(1,j).index_copy_(0,ids,terms.sum(-1));
        else dc.select(0,j).copy_(terms.sum());
      }
    }
    result[2*count] = dc;
    return result;
  }
};
} // namespace
std::vector<Tensor> isolated_aggregate(const std::vector<Tensor>& atoms,
    const std::vector<Tensor>& scales, const Tensor& coefficients, Index sources, bool mean) {
  if (sources < 1 || atoms.empty() || atoms.size()%sources || atoms.size() != scales.size())
    throw std::invalid_argument("isolated Aggregate source layout mismatch");
  const auto& first = atoms.front();
  if (!first.defined() || first.dim() != 1 || !first.device().is_cpu()
      || (first.scalar_type() != at::kFloat && first.scalar_type() != at::kDouble))
    throw std::invalid_argument("isolated Aggregate requires CPU FP32/FP64 vectors");
  for (size_t i = 0; i < atoms.size(); ++i)
    if (!atoms[i].defined() || atoms[i].sizes() != first.sizes() || atoms[i].options().dtype() != first.options().dtype()
        || atoms[i].device() != first.device() || !scales[i].defined() || scales[i].dim() != 0
        || scales[i].scalar_type() != first.scalar_type() || scales[i].device() != first.device())
      throw std::invalid_argument("isolated Aggregate input metadata mismatch");
  const bool shape = coefficients.defined() && ((coefficients.dim() == 1 && (!coefficients.numel() || coefficients.numel() == sources))
      || (coefficients.dim() == 2 && coefficients.size(0) == static_cast<Index>(atoms.size())/sources && coefficients.size(1) == sources));
  if (!shape
      || coefficients.scalar_type() != first.scalar_type() || coefficients.device() != first.device()
      || (mean && coefficients.numel())) throw std::invalid_argument("isolated Aggregate coefficients mismatch");
  return IsolatedAggregate::apply(at::TensorList(atoms), at::TensorList(scales), coefficients, sources, mean);
}
} // namespace tide

File: python/tidegraph/isolated_aggregate.py
"""Packed source arithmetic with separate summary/contribution gradient roots."""
import torch


class _Sources(torch.autograd.Function):
    @staticmethod
    def forward(ctx, sources, mean, coefficients, *inputs):
        ctx.set_materialize_grads(False)
        count = len(inputs)//2
        atoms, scales = inputs[:count], inputs[count:]
        rows = count//sources
        x = torch.stack(atoms).reshape(rows, sources, -1)
        s = torch.stack(scales).reshape(rows, sources)
        terms = x*s.unsqueeze(-1)
        if mean:
            terms = terms/sources
        elif coefficients.numel():
            terms = terms*coefficients.reshape(-1, sources, 1)
        total = terms[:, 0]
        for j in range(1, sources):
            total = total + terms[:, j]
        outputs, frozen = [], []
        for i in range(rows):
            summary = total[i]
            outputs.append(summary)
            active = False
            for j in range(sources):
                k = i*sources+j
                value = terms[i, j]
                outputs.append(value)
                needs = atoms[k].requires_grad or scales[k].requires_grad or coefficients.requires_grad
                active |= needs
                if not needs:
                    frozen.append(value)
            if not active:
                frozen.append(summary)
        ctx.mark_non_differentiable(*frozen)
        ctx.sources, ctx.mean = sources, mean
        ctx.save_for_backward(x, s, coefficients)
        return tuple(outputs)

    @staticmethod
    def backward(ctx, *bars):
        if torch.is_grad_enabled():
            raise ValueError("isolated Aggregate supports first-order VJP only")
        sources = ctx.sources
        rows = len(bars)//(sources+1)
        count = rows*sources
        result = [None]*(3+2*count)
        x, s, coe = ctx.saved_tensors
        dc = None
        for j in range(sources):
            used, values = [], []
            for i in range(rows):
                total, part = bars[i*(sources+1)], bars[i*(sources+1)+1+j]
                if total is None and part is None:
                    continue
                used.append(i)
                values.append(part if total is None else total if part is None else total+part)
            if not used:
                continue
            dy = torch.stack(values)
            xs, ss = x[used, j], s[used, j]
            dv = dy/sources if ctx.mean else dy
            if not ctx.mean and coe.numel():
                dv = dy*(coe[used, j].unsqueeze(-1) if coe.ndim == 2 else coe[j])
            if any(ctx.needs_input_grad[3+i*sources+j] for i in used):
                dx = dv*ss.unsqueeze(-1)
                for k, i in enumerate(used):
                    if ctx.needs_input_grad[3+i*sources+j]:
                        result[3+i*sources+j] = dx[k]
            if any(ctx.needs_input_grad[3+count+i*sources+j] for i in used):
                ds = (dv*xs).sum(-1)
                for k, i in enumerate(used):
                    if ctx.needs_input_grad[3+count+i*sources+j]:
                        result[3+count+i*sources+j] = ds[k]
            if coe.numel() and ctx.needs_input_grad[2]:
                if dc is None:
                    dc = torch.zeros_like(coe)
                terms = dy*(xs*ss.unsqueeze(-1))
                if coe.ndim == 2:
                    dc[used, j] = terms.sum(-1)
                else:
                    dc[j] = terms.sum()
        result[2] = dc
        return tuple(result)


def aggregate(atoms, scales, coefficients, sources, mean=False):
    if type(sources) is not int or sources < 1 or not atoms or len(atoms)%sources or len(atoms) != len(scales):
        raise ValueError("isolated Aggregate source layout mismatch")
    first = atoms[0]
    if first.ndim != 1 or first.device.type not in {"cpu", "cuda", "npu"} or first.dtype not in (torch.float32, torch.float64):
        raise ValueError("isolated Aggregate requires supported FP32/FP64 vectors")
    if any(a.shape != first.shape or s.ndim != 0 or a.dtype != first.dtype or s.dtype != first.dtype
           or a.device != first.device or s.device != first.device for a, s in zip(atoms, scales)):
        raise ValueError("isolated Aggregate input metadata mismatch")
    if (coefficients.shape not in {(0,), (sources,), (len(atoms)//sources, sources)}
            or coefficients.dtype != first.dtype or coefficients.device != first.device
            or (mean and coefficients.numel())):
        raise ValueError("isolated Aggregate coefficients mismatch")
    return list(_Sources.apply(sources, mean, coefficients, *atoms, *scales))
